src/main/java/com/hierynomus/msfscc/fileinformation/ShareInfo.java

```diff
@@ -0,0 +1,108 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hierynomus.msfscc.fileinformation;
+
+import java.math.BigInteger;
+
+import com.hierynomus.protocol.commons.buffer.Buffer;
+import com.hierynomus.protocol.commons.buffer.Buffer.BufferException;
+
+public class ShareInfo {
+
+	private final long totalAllocationUnits;
+	private final long callerAvailableAllocationUnits;
+	private final long actualAvailableAllocationUnits;
+	private final int sectorsPerAllocationUnit;
+	private final int bytesPerSector;
+	
+	private final long totalSpace;
+	private final long callerFreeSpace;
+	private final long actualFreeSpace;
+	
+	private final static BigInteger MAX_VALUE = BigInteger.valueOf(Long.MAX_VALUE);
+	
+	ShareInfo(long totalAllocationUnits, long callerAvailableAllocationUnits,
+			long actualAvailableAllocationUnits, int sectorsPerAllocationUnit, int bytesPerSector) {
+		this.totalAllocationUnits = totalAllocationUnits;
+		this.callerAvailableAllocationUnits = callerAvailableAllocationUnits;
+		this.actualAvailableAllocationUnits = actualAvailableAllocationUnits;
+		this.sectorsPerAllocationUnit = sectorsPerAllocationUnit;
+		this.bytesPerSector = bytesPerSector;
+
+		// Using BigInteger to check for overflows...
+		BigInteger bytesPerAllocationUnit = BigInteger.valueOf(sectorsPerAllocationUnit).multiply(BigInteger.valueOf(bytesPerSector));
+		assert bytesPerAllocationUnit.compareTo(MAX_VALUE) <= 0;
+		
+		BigInteger totalSpace = BigInteger.valueOf(totalAllocationUnits).multiply(bytesPerAllocationUnit);
+		assert totalSpace.compareTo(MAX_VALUE) <= 0;
+		
+		BigInteger callerFreeSpace = BigInteger.valueOf(callerAvailableAllocationUnits).multiply(bytesPerAllocationUnit);
+		assert callerFreeSpace.compareTo(MAX_VALUE) <= 0;
+		
+		BigInteger actualFreeSpace = BigInteger.valueOf(actualAvailableAllocationUnits).multiply(bytesPerAllocationUnit);
+		assert actualFreeSpace.compareTo(MAX_VALUE) <= 0;
+		
+		this.totalSpace = totalSpace.longValue();
+		this.callerFreeSpace = callerFreeSpace.longValue();
+		this.actualFreeSpace = actualFreeSpace.longValue();
+	}
+	
+	public long getFreeSpace() {
+		return actualFreeSpace;
+	}
+	
+	public long getCallerFreeSpace() {
+		return callerFreeSpace;
+	}
+	
+	public long getTotalSpace() {
+		return totalSpace;
+	}
+	
+	public long getTotalAllocationUnits() {
+		return totalAllocationUnits;
+	}
+	
+	public long getAvailableAllocationUnits() {
+		return actualAvailableAllocationUnits;
+	}
+	
+	public long getCallerAvailableAllocationUnits() {
+		return callerAvailableAllocationUnits;
+	}
+	
+	public int getSectorsPerAllocationUnit() {
+		return sectorsPerAllocationUnit;
+	}
+	
+	public int getBytesPerSector() {
+		return bytesPerSector;
+	}
+	
+	public static ShareInfo parseFsFullSizeInformation(Buffer.PlainBuffer response) throws BufferException {
+		long totalAllocationUnits = response.readUInt64();

```

Comment:
Please add the comments to the lines, see the other buffer reading code in messages and other objects that are deserialized. This ensures that we can easily verify whether everything is correct and it makes it easy to search for